File: app/src/main/java/moe/feng/nhentai/model/Book.java
package moe.feng.nhentai.model;

import android.content.Context;
import android.util.Log;
import com.google.gson.Gson;
import java.util.ArrayList;
import moe.feng.nhentai.api.BookApi;

public class Book {

	public String title, other, bookId;

	public String previewImageUrl, bigCoverImageUrl, titleJP, titlePretty, galleryId;
	public int pageCount;

	public int thumbHeight = 0, thumbWidth = 0;

	public String parodies, parodiesID, language, group, groupID;

	public ArrayList<String> tags = new ArrayList<>();
	public ArrayList<String> tagID = new ArrayList<>();

	public ArrayList<String> characters = new ArrayList<>();
	public ArrayList<String> charactersID = new ArrayList<>();

	public ArrayList<String> artists = new ArrayList<>();
	public ArrayList<String> artistsID = new ArrayList<>();

    public String langField = LANG_UNKNOWN;
    public static final String LANG_CN = "29963", LANG_JP = "6346", LANG_GB = "12227", LANG_UNKNOWN = "0";

	public String uploadTime, uploadTimeText;

	public Book() {
		this(null, null, null);
	}

	public Book(String title, String other, String bookId) {
		this.title = title;
		this.other = other;
		this.bookId = bookId;
	}

	public Book(String title, String other, String bookId, String previewImageUrl) {
		this.title = title;
		this.other = other;
		this.bookId = bookId;
		this.previewImageUrl = previewImageUrl;
	}

	public String getAvailableTitle() {

		if (titlePretty== null) {
			Log.d(Book.class.getSimpleName(), "getAvailableTitle: Pretty Title Not Available");
		}

		else if(!titlePretty.equals("")) {
			return titlePretty;
		}

		if( (this.langField.equals(Book.LANG_JP) || this.langField.equals(Book.LANG_CN))){
			if(titleJP != null){
				if(!titlePretty.equals("")){
					return titleJP;
				}
				else{
					return this.title;
				}
			}
			else {
				return this.title;
			}
		}
			return this.title != null ? this.title : this.titleJP;
	}

	public String toJSONString() {
		return new Gson().toJson(this);
	}

	public void updateDataFromOldData(Context context) {
		Log.d("HEY", "updateDataFromOldData: " + this.bookId);
		BaseMessage msg = BookApi.getBook(context,this);
		if (msg.getCode()==0){
			Book result = msg.getData();
			this.title = result.title;
			this.titleJP = result.titleJP;
			this.titlePretty = result.titlePretty;
			this.characters = result.characters;
			this.charactersID = result.charactersID;
			this.parodies = result.parodies;
			this.parodiesID = result.parodiesID;
			this.artists = result.artists;
			this.uploadTime = result.uploadTime;
			this.uploadTimeText = result.uploadTimeText;
			this.artistsID =result.artistsID;
			this.bigCoverImageUrl = result.bigCoverImageUrl;
			this.langField = result.langField;
			this.language = result.language;
			this.tags = result.tags;
			this.tagID = result.tagID;
			this.galleryId = result.galleryId;
			this.bookId = result.bookId;
			this.group = result.group;
			this.groupID = result.groupID;
			this.previewImageUrl = result.previewImageUrl;
			this.other =result.other;
		}
	}

	public static Book toBookFromJson(String json) {
		return  new Gson().fromJson(json, Book.class);
	}

}

File: app/src/main/kotlin/moe/feng/nhentai/model/Book.kt
package moe.feng.nhentai.model

import android.arch.persistence.room.Entity
import android.arch.persistence.room.PrimaryKey
import com.google.gson.annotations.SerializedName
import moe.feng.nhentai.R
import moe.feng.nhentai.api.ApiConstants
import java.text.DateFormat
import java.util.*

@Entity(tableName = Book.TAG) class Book: History.KeyContainer<Book.HistoryKey> {

	// Serializable fields
	@PrimaryKey @SerializedName("id") lateinit var bookId: String
	@SerializedName("media_id") lateinit var galleryId: String
	private @SerializedName("title") var titles: BookTitle = BookTitle()
	@SerializedName("upload_date") var uploadDate: Long = 0L
	@SerializedName("num_favorites") var favCount: Int = 0
	var scanlator: String = ""
	@SerializedName("num_pages") var pageCount: Int = 0
	var images: BookImages = BookImages()
	@SerializedName("tags") var allTags: Array<Tag> = emptyArray()

	// Favourite mark
	var isFavourite: Boolean = false

	// Methods
	val language: String get() = allTags.find {
		it.type == Tag.TYPE_LANGUAGE && it.name != Tag.LANG_TRANSLATED }?.name ?: Tag.LANG_JAPANESE
	val tags: Array<Tag> get() = allTags.filter { it.type == Tag.TYPE_TAG }.toTypedArray()
	val categories: Array<Tag> get() = allTags.filter { it.type == Tag.TYPE_CATEGORY }.toTypedArray()
	val characters: Array<Tag> get() = allTags.filter { it.type == Tag.TYPE_CHARACTER }.toTypedArray()
	val groups: Array<Tag> get() = allTags.filter { it.type == Tag.TYPE_GROUP }.toTypedArray()
	val parodies: Array<Tag> get() = allTags.filter { it.type == Tag.TYPE_PARODY }.toTypedArray()
	val artists: Array<Tag> get() = allTags.filter { it.type == Tag.TYPE_ARTIST }.toTypedArray()

	val title: String get() = titles.english
	val jpTitle: String? get() = titles.japanese
	val prettyTitle: String? get() = titles.pretty

	fun getLanguageFlagRes(): Int = when (language) {
		Tag.LANG_CHINESE -> R.drawable.ic_lang_cn
		Tag.LANG_ENGLISH -> R.drawable.ic_lang_gb
		Tag.LANG_JAPANESE -> R.drawable.ic_lang_jp
		else -> 0
	}

	fun getFormattedTime(): String = DateFormat.getDateTimeInstance().format(Date(uploadDate * 1000))

	override fun getHistoryAction(): Int = History.ACTION_READ_BOOK

	override fun getHistoryKey(): Book.HistoryKey
			= HistoryKey(bookId, galleryId, titles, images.cover)

	// Sub-classes

	data class BookTitle(
			var japanese: String? = null,
			var pretty: String? = null,
			var english: String = ""
	)

	class BookImages(
			var cover: Picture? = null,
	        var pages: Array<Picture> = emptyArray(),
	        var thumbnail: Picture? = null
	)

	class HistoryKey(
			val id: String,
	        val galleryId: String,
	        val titles: BookTitle,
	        val cover: Picture? = null
	): History.Key {
		override fun id(): String = id
	}

	companion object {

		const val TAG = "Book"

	}

	// Extend function

	val requestUrl: String get() = ApiConstants.getBookDetailsUrl(bookId)
	val thumbUrl: String get() = ApiConstants.getBookThumbUrl(galleryId, images.cover?.fileType)
	val bigCoverUrl: String get() = ApiConstants.getBigCoverUrl(galleryId)
	val pagePictures: BookPageGetter get() = BookPageGetter(this, false)
	val pageThumbnails: BookPageGetter get() = BookPageGetter(this, true)

	class BookPageGetter internal constructor(private val book: Book, private val thumb: Boolean) {

		operator fun get(pageNum: Int): String =
				(if (thumb) ApiConstants::getPictureUrl else ApiConstants::getThumbPictureUrl)(
						book.galleryId,
						pageNum.toString(),
						book.images.pages[pageNum].fileType
				)

	}

}